### PR TITLE
[ws-manager] Switch to controller-runtime abstraction

### LINKED ISSUE #2877: [ws-manager] Use SharedInformer instead of directly asking the data plane
At the moment ws-manager always, directly talks to the Kubernetes API if it wants information about a pod, service or configmap. Instead, we should use a [SharedInformer](https://pkg.go.dev/k8s.io/client-go/informers) 
- to ease the load on the Kubernetes master. This might help us prevent/mitigate issues like #2870.
- to be more in line with Kubernetes controller best practice.

Potential pitfalls:
- in ws-manager after we've modified an object, we assume that a subsequent `Get` will return this modified object. Informers may break this assumption because they might not have incorporated the update yet. We might have to explicitly sync here.

### PR DESCRIPTION
**Pull request by aledbf**
Alternative implementation to https://github.com/gitpod-io/gitpod/pull/3208

Tasks:

- Use controller-runtime instead of direct calls to the API server. Operations Get and List use a local cache
- Remove watch of pods and configmaps using ResourceEventHandler to enqueue events
- Unit tests use [envtest](https://book.kubebuilder.io/reference/envtest.html) package from [controller-runtime](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/envtest)
  (this approach ensures the data used in the tests pass k8s validations)

A full example of the new metrics available: https://gist.github.com/aledbf/64ecec50c8d8bc95a345437cf3561a22

fixes #2877

replaces #3229
replaces #3227
replaces #3226 
replaces #3208

cc @csweichel

### CODE REVIEW

File: components/ws-manager/BUILD.yaml [hunk 1]
```diff
@@ -63,3 +64,8 @@ packages:
         # - ["sh", "-c", "./components-ws-manager--app/ws-manager --kubeconfig kubeconfig integration-test objs chart--helm/gitpod | kubectl --kubeconfig kubeconfig apply -f -"]
         # - ["components-ws-manager--app/_tests/manager.test", "-integration-test", "kubeconfig", "-test.run", "TestIntegration", "-test.v"]
         # - ["kubectl", "--kubeconfig", "kubeconfig", "delete", "namespace", "wsman-integration-${__pkg_version}"]
+  - name: kubebuilder-binaries
```
Comment (aledbf) [inline components/ws-manager/BUILD.yaml:67]: @csweichel how I can disable cache for this task? Or better yet, how I can add these binaries to the image build image
(I tried updating the `dev-environment` image without success. Maybe this needs to be done in a different PR?)

Comment (csweichel) [reply components/ws-manager/BUILD.yaml:67]: You can add them to the dev-environment [Dockerfile](https://github.com/gitpod-io/gitpod/blob/master/dev/image/Dockerfile), let werft build the image for you (see werft result page) and then use that image in the `.gitpod.yml` and `.werft/build.yaml`.

**Comment by ArthurSens** [comment]
> A full example of the new metrics available: https://gist.github.com/aledbf/64ecec50c8d8bc95a345437cf3561a22

This seems to be way too expensive :grimacing: 
I'm afraid that we could OOM kill Prometheus with this amount of metrics. We can always drop them at scrape time, but we'd need to make some config changes on Prometheus' side for that to happen.

File: components/ws-manager/pkg/manager/manager.go [hunk 2]
```diff
@@ -1025,22 +1055,23 @@ func (m *Manager) GetWorkspaces(ctx context.Context, req *api.GetWorkspacesReque
 // getAllWorkspaceObjects retturns all (possibly incomplete) workspaceObjects of all workspaces this manager is currently aware of.
 // If a workspace has a pod that pod is part of the returned WSO.
 // If a workspace has a PLIS that PLIS is part of the returned WSO.
-func (m *Manager) getAllWorkspaceObjects(ctx context.Context) (result []workspaceObjects, err error) {
+func (m *Manager) getAllWorkspaceObjects(ctx context.Context) ([]workspaceObjects, error) {
 	//nolint:ineffassign
 	span, ctx := tracing.FromContext(ctx, "getAllWorkspaceObjects")
-	defer tracing.FinishSpan(span, &err)
+	defer tracing.FinishSpan(span, nil)
```
Comment (csweichel) [inline components/ws-manager/pkg/manager/manager.go:1061]: see comment above re. `FinishSpan`

File: components/ws-manager/pkg/manager/pod_controller.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// PodReconciler reconciles a Pod object
+type PodReconciler struct {
```
Comment (csweichel) [inline components/ws-manager/pkg/manager/pod_controller.go:33]: While reading through I was left wondering why a `_controller.go` containers a `*Reconciler` only. Is this a pattern in Kubernetes controller?

Personally, I'd probably throw the `ConfigmapReconciler` and `PodReconciler` into a `reconciler.go` - but that's nothing more than personal taste :)

File: components/ws-manager/go.mod [hunk 4]
```diff
@@ -53,52 +55,52 @@ replace github.com/gitpod-io/gitpod/ws-daemon/api => ../ws-daemon-api/go // leew
 
 replace github.com/gitpod-io/gitpod/ws-manager/api => ../ws-manager-api/go // leeway
 
-replace k8s.io/api => k8s.io/api v0.20.1 // leeway indirect from components/common-go:lib
+replace k8s.io/api => k8s.io/api v0.20.4 // leeway indirect from components/common-go:lib
```
Comment (csweichel) [inline components/ws-manager/go.mod:58]: we're trying to keep the Kubernetes dependencies consistent across the project. To that end there's a [`go-get-kubernetes.sh`](https://github.com/gitpod-io/gitpod/blob/b7dda84ad0c0dd195b829b7f4160f9f7acfa355c/components/common-go/go-get-kubernetes.sh) script, which in combination with `leeway link` updates Kubernetes dependencies.

I realise this is missing in the "How we develop" page and have updated [that page accordingly](https://www.notion.so/gitpod/How-we-develop-e8ce7e562478458a9223eb9b797415b2#f713c90fb99e4a9791232ceb646a15fb).

**Review by csweichel — COMMENTED** [review]
Wonderful change. Thank you so much for moving ws-manager closer to Kubernetes - and making it easier on the data plane.

File: components/ws-manager/BUILD.yaml (hunk 1 above)
Comment (csweichel) [reply components/ws-manager/BUILD.yaml]: After going through the code I was left wondering what we'd be using `kubebuilder` for. Was it used for the metrics that were previously part of this PR?

File: components/ws-manager/cmd/run.go [hunk 5]
```diff
@@ -44,23 +51,64 @@ var runCmd = &cobra.Command{
...
-		clientset, err := newClientSet()
+		ctrl.SetLogger(logrusr.NewLogger(log.Log))
+
+		opts := ctrl.Options{
+			Scheme:    scheme,
+			Namespace: cfg.Manager.Namespace,
+			//Port:                   9443,
+			HealthProbeBindAddress: ":0",
+			LeaderElection:         false,
+			LeaderElectionID:       "706fe79e.gitpod.io",
```
[the remark this answers is not in the document]
Comment (aledbf) [reply components/ws-manager/cmd/run.go:62]: Is an arbitrary unique string (any suggestion is welcome)

File: components/ws-manager/BUILD.yaml (hunk 1 above)
Comment (aledbf) [reply components/ws-manager/BUILD.yaml]: Is for [envtest](https://book.kubebuilder.io/reference/envtest.html) The binaries are related to kube-apiserver and etcd

File: components/ws-manager/pkg/manager/configmap_controller.go [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+	"k8s.io/apimachinery/pkg/watch"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// ConfigmapReconciler reconciles a Configmap object
+type ConfigmapReconciler struct {
+	client.Client
+	Log    logr.Logger
```
[the remark this answers is not in the document]
Comment (aledbf) [reply components/ws-manager/pkg/manager/configmap_controller.go:35]: Most of the code in k8s uses [klog](https://github.com/kubernetes/klog) (including client-go). That library now uses [logr](github.com/go-logr/logr) underneath. For this reason, controller-runtime, kubebuilder, and similar use logr as log interface.
The dependency I added in this PR is just a logrus bridge to logr.

File: components/ws-manager/cmd/run.go [hunk 7]
```diff
@@ -114,47 +162,57 @@ var runCmd = &cobra.Command{
...
+				log.WithError(err).Fatal("cannot start workspace monitor")
+			}
+		}()
+
 		defer monitor.Stop()
 		log.Info("workspace monitor is up and running")
 
-		if cfg.PProf.Addr != "" {
-			go pprof.Serve(cfg.PProf.Addr)
+		if err = (&manager.ConfigmapReconciler{
```
[the remark this answers is not in the document]
Comment (aledbf) [reply components/ws-manager/cmd/run.go]: fixed

File: components/ws-manager/pkg/manager/manager.go [hunk 8]
```diff
@@ -584,10 +615,10 @@ func (m *Manager) markAllWorkspacesActive() error {
 }
 
 // ControlPort publicly exposes or un-exposes a network port for a workspace
-func (m *Manager) ControlPort(ctx context.Context, req *api.ControlPortRequest) (res *api.ControlPortResponse, err error) {
+func (m *Manager) ControlPort(ctx context.Context, req *api.ControlPortRequest) (*api.ControlPortResponse, error) {
 	span, ctx := tracing.FromContext(ctx, "ControlPort")
 	tracing.ApplyOWI(span, log.OWI("", "", req.Id))
-	defer tracing.FinishSpan(span, &err)
+	defer tracing.FinishSpan(span, nil)
```
[the remark this answers is not in the document]
Comment (aledbf) [reply components/ws-manager/pkg/manager/manager.go]: fixed

File: components/ws-manager/pkg/manager/manager.go (hunk 2 above)
Comment (aledbf) [reply components/ws-manager/pkg/manager/manager.go:1061]: fixed

File: components/ws-manager/go.mod (hunk 4 above)
Comment (aledbf) [reply components/ws-manager/go.mod]: fixed

File: components/ws-manager/pkg/manager/manager.go [hunk 9]
```diff
@@ -761,14 +791,14 @@ func (m *Manager) ControlPort(ctx context.Context, req *api.ControlPortRequest)
...
-		_, err = client.Services(m.Config.Namespace).Update(ctx, service, metav1.UpdateOptions{})
+		err = m.Clientset.Update(ctx, &service)
 		if err != nil {
 			return nil, xerrors.Errorf("cannot update service: %w", err)
 		}
 		tracing.LogEvent(span, "port service updated")
 	}
 	if err != nil {
-		return nil, xerrors.Errorf("cannot control port: %w", err)
+		return nil, xerrors.Errorf("cannot control port (2): %w", err)
```
[the remark this answers is not in the document]
Comment (aledbf) [reply components/ws-manager/pkg/manager/manager.go]: fixed

File: components/ws-manager/pkg/manager/create.go [hunk 10]
```diff
@@ -646,7 +647,7 @@ func (m *Manager) createDefaultSecurityContext() (*corev1.SecurityContext, error
 	return res, nil
 }
 
-func (m *Manager) createPortsService(workspaceID string, metaID string, servicePrefix string, ports []*api.PortSpec) (*corev1.Service, error) {
+func (m *Manager) createPortsService(workspaceID string, metaID string, servicePrefix string, ports []*api.PortSpec) (corev1.Service, error) {
```
[the remark this answers is not in the document]
Comment (aledbf) [reply components/ws-manager/pkg/manager/create.go]: fixed

File: components/ws-manager/pkg/manager/integration_test.go [hunk 11]
```diff
@@ -195,13 +226,23 @@ func (r *StatusRecoder) Log() []api.WorkspaceStatus {
...
+func ensureIntegrationTestTheiaLabelOnNodes(clientset client.Client, namespace string) (version string, err error) {
 	version = "wsman-test"
 
 	ctx, cancel := context.WithTimeout(context.Background(), kubernetesOperationTimeout)
 	defer cancel()
 
-	nodes, err := client.CoreV1().Nodes().List(ctx, metav1.ListOptions{LabelSelector: fmt.Sprintf("gitpod.io/theia.%s", version)})
+	selector, err := labels.Parse(fmt.Sprintf("gitpod.io/theia.%s", version))
+	if err != nil {
+		return "", fmt.Errorf("cannot build label selector for theia: %v", err)
```
[the remark this answers is not in the document]
Comment (aledbf) [reply components/ws-manager/pkg/manager/integration_test.go]: fixed

File: components/ws-manager/pkg/manager/manager_test.go [hunk 12]
```diff
@@ -73,39 +74,49 @@ func TestControlPort(t *testing.T) {
...
 				return &result
 			}
 
-			ctx, cancel := context.WithTimeout(context.Background(), kubernetesOperationTimeout)
-			defer cancel()
-
 			result.Response = resp
-			result.PortsService, _ = manager.Clientset.CoreV1().Services(manager.Config.Namespace).Get(ctx, getPortsServiceName(startCtx.Request.ServicePrefix), metav1.GetOptions{})
+
+			time.Sleep(500 * time.Millisecond)
```
[the remark this answers is not in the document]
Comment (aledbf) [reply components/ws-manager/pkg/manager/manager_test.go:112]: done

File: components/ws-manager/pkg/manager/configmap_controller.go [hunk 13]
```diff
@@ -0,0 +1,72 @@
+/*
```
[the remark this answers is not in the document]
Comment (aledbf) [reply components/ws-manager/pkg/manager/configmap_controller.go:1]: I get no changes after I run the command:
```
❯ leeway run components:update-license-header
📦  cached  dev/addlicense:app                     (version 1e98a176bd46cba255521e2548dbc5a77ec4dec2)
🔧  build   components:update-license-header-deps  (version 888ced048370b237a92dfb04a050ab35d2461920)

[components:update-license-header-deps] build started (version 888ced048370b237a92dfb04a050ab35d2461920)
[components:update-license-header-deps] package build succeded (0.00s)

build succeded
Temp Dir: /tmp/tmp.MY5CkqO2Yr
FATA[0000] open WORKSPACE.yaml: no such file or directory 
FATA[0000] open WORKSPACE.yaml: no such file or directory
```

File: components/ws-manager/pkg/manager/manager_test.go [hunk 14]
```diff
@@ -155,6 +198,25 @@ func TestGetWorkspaces(t *testing.T) {
 	test.Run()
 }
 
+func removeTemporalAttributes(workspaceStatus []*api.WorkspaceStatus) []*api.WorkspaceStatus {
```
[the remark this answers is not in the document]
Comment (aledbf) [reply components/ws-manager/pkg/manager/manager_test.go:202]: done

File: components/ws-manager/cmd/run.go [hunk 15]
```diff
@@ -114,47 +162,57 @@ var runCmd = &cobra.Command{
...
 			log.WithError(err).Fatal("cannot start workspace monitor")
 		}
-		err = monitor.Start()
-		if err != nil {
-			log.WithError(err).Fatal("cannot start workspace monitor")
-		}
+
+		go func() {
+			// Monitor need to access local cache
+			time.Sleep(10 * time.Second)
```
[the remark this answers is not in the document]
Comment (aledbf) [reply components/ws-manager/cmd/run.go]: done

File: components/ws-manager/pkg/manager/configmap_controller.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (csweichel) [reply components/ws-manager/pkg/manager/configmap_controller.go:1]: You'd have to remove the apache license header first, as the license updater detects its presence and doesn't do anything.

File: components/ws-manager/pkg/manager/pod_controller.go (hunk 3 above)
Comment (aledbf) [reply components/ws-manager/pkg/manager/pod_controller.go:21]: >  Is this a pattern in Kubernetes controller?

Yes, but also the _controller.go file should contain the logic/helpers in the *Reconciler method. I see this PR as the first step to refactor the code in a different PR.

**Review by csweichel — APPROVED** [review]
Awesome stuff. Thank you so much for bringing the codebase more in-line with Kubernetes.

Prior to merging, please squash/author the commits for a clean history.

**Pull request merged**